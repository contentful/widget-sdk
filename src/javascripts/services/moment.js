'use strict';

angular.module('cf.utils')
.factory('moment', ['raw/moment', function (moment) {
  moment.locale('en', {
    calendar: {
      lastDay : '[Yesterday], LT',
      sameDay : '[Today], LT',
      nextDay : '[Tomorrow], LT',
      lastWeek : 'ddd, LT',
      nextWeek : '[Next] ddd, LT',
      sameElse : 'll'
    }
  });

  return moment;
}]);
