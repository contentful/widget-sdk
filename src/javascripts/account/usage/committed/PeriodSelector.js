import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';

import { periodPropType } from './propTypes.es6';

const formatDate = date => moment(date).format('DD MMM');

export default class PeriodSelector extends React.Component {
  static propTypes = {
    periods: PropTypes.arrayOf(periodPropType).isRequired,
    selectedPeriodIndex: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
  };

  render() {
    const { periods, onChange, selectedPeriodIndex } = this.props;
    return (
      <span className="usage__period-selector">
        <label>Usage period</label>
        <Select
          defaultValue={selectedPeriodIndex}
          onChange={onChange}
          width="auto"
          name="period-selector"
          id="period-selector">
          {periods.map(({ startDate, endDate }, index) => {
            const isCurrentPeriod = endDate === null;
            const start = moment(startDate);
            const end = isCurrentPeriod
              ? moment(start)
                  .add(1, 'month')
                  .subtract(1, 'day')
              : moment(endDate);

            const value = `${index}`;

            return (
              <Option key={index} value={value}>{`${formatDate(start)} â€“ ${formatDate(end)} ${cond([
                [constant(isCurrentPeriod), constant('(current)')],
                [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
                [constant(stubTrue), constant(end.year())]
              ])()}`}</Option>
            );
          })}
        </Select>
      </span>
    );
  }
}
