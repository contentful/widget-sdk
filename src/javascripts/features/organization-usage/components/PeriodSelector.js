import React from 'react';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';
import { track } from 'analytics/Analytics';
import { useUsageState, useUsageDispatch } from '../hooks/usageContext';
import { pick } from 'lodash/fp';
import { loadPeriodData } from '../services/UsageService';

const formatDate = (date) => moment(date).format('DD MMM');

export const PeriodSelector = () => {
  const { periods, selectedPeriodIndex, orgId, isTeamOrEnterpriseCustomer } = useUsageState();
  const dispatch = useUsageDispatch();

  const handleChange = async (e) => {
    dispatch({ type: 'SET_LOADING', value: true });

    const newPeriodIndex = parseInt(e.target.value);

    // analytics
    const oldPeriod = periods[selectedPeriodIndex];
    const newPeriod = periods[newPeriodIndex];
    track('usage:period_selected', {
      oldPeriod: pick(['startDate', 'endDate'], oldPeriod),
      newPeriod: pick(['startDate', 'endDate'], newPeriod),
    });

    dispatch({ type: 'CHANGE_PERIOD', value: newPeriodIndex });
    dispatch({ type: 'SWITCH_MAIN_TAB', value: 'apiRequest' });
    const data = await loadPeriodData(orgId, newPeriod);
    dispatch({ type: 'SET_USAGE_DATA', value: data });
    dispatch({ type: 'SET_LOADING', value: false });
  };

  const processDate = (startDate, endDate) => {
    const isCurrentPeriod = endDate === null || moment().diff(moment(endDate), 'days') == 0; // the end day is today
    const start = moment(startDate);
    const end = isCurrentPeriod
      ? moment(start).add(1, 'month').subtract(1, 'day')
      : moment(endDate);
    return `${formatDate(start)} â€“ ${formatDate(end)} ${cond([
      [constant(isCurrentPeriod), constant('(current)')],
      [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
      [constant(stubTrue), constant(end.year())],
    ])()}`;
  };

  const renderSelect = () => {
    return (
      <Select
        defaultValue={selectedPeriodIndex}
        onChange={handleChange}
        width="auto"
        name="period-selector"
        testId="period-selector"
        id="period-selector">
        {periods.map(({ startDate, endDate }, index) => {
          const value = `${index}`;

          return (
            <Option key={index} value={value}>
              {processDate(startDate, endDate)}
            </Option>
          );
        })}
      </Select>
    );
  };

  const renderCurrentPeriodTextOnly = () => {
    const { startDate, endDate } = periods[0];
    return <strong data-test-id="usage-period-text">{processDate(startDate, endDate)}</strong>;
  };

  return (
    <div className="usage__period-selector">
      <label>API requests usage period</label>
      {isTeamOrEnterpriseCustomer ? renderSelect() : renderCurrentPeriodTextOnly()}
    </div>
  );
};
