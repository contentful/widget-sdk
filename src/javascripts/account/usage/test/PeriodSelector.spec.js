import { shallow } from 'enzyme';
import React from 'react';
import moment from 'moment';
import { Select } from '@contentful/ui-component-library';

import PeriodSelector from '../committed/PeriodSelector.es6';

let wrapper = null;
let props = null;
describe('PeriodSelector', () => {
  beforeEach(() => {
    const startDate = moment().subtract(12, 'days');
    props = {
      periods: [
        {
          sys: { type: 'UsagePeriod', id: 'period1' },
          startDate: startDate.toISOString(),
          endDate: null
        },
        {
          sys: { type: 'UsagePeriod', id: 'period1' },
          startDate: moment(startDate)
            .subtract(1, 'day')
            .subtract(1, 'month')
            .toISOString(),
          endDate: moment(startDate)
            .subtract(1, 'day')
            .toISOString()
        }
      ],
      selectedPeriodIndex: 0,
      onChange: jest.fn()
    };
    wrapper = shallow(<PeriodSelector {...props} />);
  });

  it('should call onChange when select was used', () => {
    expect(props.onChange).not.toHaveBeenCalled();
    wrapper.find(Select).simulate('change', { target: { value: 1 } });
    expect(props.onChange).toHaveBeenCalledWith({ target: { value: 1 } });
  });
});
