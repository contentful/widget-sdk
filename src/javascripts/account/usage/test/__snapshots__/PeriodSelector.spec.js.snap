// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeriodSelector should render 1`] = `
<span
  className="usage__period-selector"
>
  <label>
    Usage period
  </label>
  <Select
    defaultValue={0}
    hasError={false}
    id="period-selector"
    isDisabled={false}
    name="period-selector"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onFocus={[Function]}
    required={false}
    testId="cf-ui-select"
    width="auto"
  >
    <Option
      key="0"
      value="0"
    >
      10 Oct – 09 Nov (current)
    </Option>
    <Option
      key="1"
      value="1"
    >
      09 Sep – 09 Oct (a month ago)
    </Option>
  </Select>
</span>
`;
